### ISSUE TITLE
Help using "feature/diagnostic-tools" in a test environment

### ISSUE THREAD

**Issue by timezombi** [Unity-Technologies/unity-cache-server#135]
As discussed in #91 and #114, I am setting up a test environment with production level resources in order to stress test 6.4.0b2.

I pulled from `feature/diagnostic-tools` to see what tools were available to assist with stress testing. I'm still waiting on a maintenance window to switch over some build nodes to the test environment, but I have UCS running with the `--diag-client-recorder` parameter and default settings.

In the README, there was mention of a `ucs-diag-tools` repository where the diagnostic utilities will go. However, that repo is empty. The old `stream_player.js` was deleted from the `feature/diagnostic-tools` directory.

Is the `stream_player.js` utility going to be added to `ucs-diag-tools` soon? Would there be any problem if I just added the `stream_player.js` file back into my local repo?

If I manage to consistently crash UCS, would sending the `client-recordings` help you troubleshoot the issue?

Thank you in advance for helping me test this!

**Comment by stephen-palmer** [comment]
Hi!

I just pushed/published a bunch of things to make this easier, including the promised ucs-diag-utils repo.

1) install beta 3 of the cache server, which now has the `--diag-client-recorder` functionality.
2) install beta of [ucs-diag-utils](https://github.com/Unity-Technologies/ucs-diag-utils)  `npm install -g ucs-diag-utils@beta` or use `stream_player.js` in the github repo directly.
3) after recording a client session, you can easily debug a stream: `ucs-stream-player -d <stream-path>` or re-play it to a cache server: `ucs-stream-player <stream-path]> <server:port>`

Let me know if you have any trouble with that. And yes, if you have a stream file that can repro a crash, I'd love to have it.

**Comment by timezombi** [comment]
I'm reaching the tail end of my testing. So far things have been looking extremely promising. I only have a couple more tests to do.

Since there doesn't appear to be a real-time progress indicator for the stream_player, I'm not sure if it hangs on large client recordings or if it is just going to take a really long time. I'll let it run overnight. It is currently attempting to replay the largest client recording I have (50GB) 6 times concurrently.

If it completes overnight, I'll run through another round of huge transaction imports to see if it stalls Unity Editors that are trying to connect.

It it passes those tests, it is going into production this weekend.

I'll follow up again after the final tests are done.

**Comment by stephen-palmer** [comment]
Encouraging reports Brandon, thanks for the updates.

I am pretty confident `stream_player` isn't hanging - it's just waiting for async I/O to finish. You are testing at quite an extreme scale, putting Node.js to the test as much as the cache server code!

**Comment by timezombi** [comment]
I finished all the stress testing that I will be doing. The diagnostic tools never did finish replaying my largest client recording, but that is not a deal breaker. The parameters that were breaking the diagnostic tools and cache server were unlikely or impossible to be encountered in real-world usage.

The results of the testing gave me enough confidence to deploy 6.4.0b3 into production. I was not able to repro _any_ OOM errors despite simulating workloads far greater than what we could ever realistically produce in production.

It seems like 6.4.0 scales much better than 6.3.0 when handling simultaneous large transactions. With ludicrously large simulated workloads, memory usage did not go over 2.2GB total. On 6.3.0, we would run out of memory (48GB) with comparatively modest workloads.

I'll admit my testing for idle memory usage for 6.3.0 wasn't particularly rigorous, but in general 6.4.0  seemed to use up to 50% less memory than 6.3.0 when idling. Transfer speeds were also not noticeably different than 6.3.0.

During testing, I was able to consistently crash the cache server (causing indefinite hangs) by making many rapid connections to the cache server. On the first day of launching the cache server, our graphics programmer unintentionally did something to generate ~1700 connections within about an hour, ~700 of which were made within the last 10 minutes. This seems to have silently crashed the cache server. There was no noteworthy errors in the logs, but the process did exit (instead of hanging indefinitely), which caused my restart script to spin up a new cache server instance.

Our graphics programmer crashing the cache server did worry me a bit, but it turns out he did something really unusual to cause so many connections. I'm not sure if I understood correctly, but I think he said he mistakenly did a revert of thousands of files while simultaneously making a build, a scenario which should not happen again. Aside from this isolated incident, Light Baking seems to generate the most connections, but not at the scale needed to crash the server.

Since I suspect the workloads that have crashed this cache server would have also crashed 6.3.0, I'm so far happy with the results.

So far it has just crashed the one time while deployed in production. I will report back again after running it for a while longer.
